docs/options.md
---
title: Options
permalink: /options
---
There are several options available to customize the `Server` behavior.
These are specifiable using the `Options` class and passing such an instance to the `Server` constructor.

## Common Options

{% include undocumented.md %}


docs/classes/options.md
---
title: Options
permalink: /classes/options
---

{:.warning}
> This page needs an update.

* Table of Contents
{:toc}

The `Options` class exposes no methods, just properties. The properties may only be set during `Server` startup, after that, they're locked from further writes.

## `$debug`

Indicates whether debug mode is active.

Type: boolean

## `$user`

Only relevant under *nix systems when the server is started as root; it indicates to which user the server will switch to after startup.

Type: string &mdash; Default: current user

## `$maxConnections`

Maximum number of total simultaneous connections the server accepts. If that number is exceeded, new connections are dropped.

Type: integer greater than 0 &mdash; Default: `1000`

## `$connectionsPerIP`

Maximum number of allowed connections from an individual /32 IPv4 or /56 IPv6 range.

Type: integer greater than 0 &mdash; Default: `30`

## `$maxRequestsPerConnection`

Maximum number of requests on a single connection. (On HTTP/1.1 it controls the Keep-Alive header.)

Set to `PHP_INT_MAX` to effectively disable this limit.

Type: integer greater than 0 &mdash; Default: `1000`

## `$connectionTimeout`

Time in seconds until a connection is closed (if no further data comes in).

Type: integer greater than 0 &mdash; Default: `6`

## `$defaultContentType`

Content type of responses, if not otherwise specified by the request handler.

Type: string &mdash; Default: `"text/html"`

## `$defaultTextCharset`

Text charset of text/ content types, if not otherwise specified by the request handler.

Type: string &mdash; Default: `"utf-8"`

## `$sendServerToken`

Whether a `Server` header field should be sent along with the response.

Type: boolean &mdash; Default: `false`

## `$socketBacklogSize`

Size of the backlog, i.e. how many connections may be pending in an unaccepted state.

Type: integer greater than or equal to 16 &mdash; Default: `128

## `$normalizeMethodCase`

Whether method names shall be lowercased.

Type: boolean &mdash; Default: `true`

## `$maxConcurrentStreams`

Maximum number of concurrent HTTP/2 streams per connection.

Type: integer greater than 0 &mdash; Default: `20`

## `$maxFramesPerSecond`

Maximum number of frames a HTTP/2 client is allowed to send per second.

Type: integer &mdash; Default: `60`

## `$allowedMethods`

Array of allowed HTTP methods. [The [`Router`](router.md) class will extend this array with the used methods.]

Type: array&lt;string> &mdash; Default: `["GET", "POST", "PUT", "PATCH", "HEAD", "OPTIONS", "DELETE"]`

## `$deflateEnable`

Whether HTTP body compression should be active.

Type: boolean &mdash; Default: `extension_loaded("zlib")`

## `$deflateContentTypes`

A regular expression to match the content-type against in order to determine whether a request shall be deflated or not.

Type: string &mdash; Default: `"#^(?:text/.*+|[^/]*+/xml|[^+]*\+xml|application/(?:json|(?:x-)?javascript))$#i"`

## `$configPath`

Path to the used configuration file

Type: string

## `$maxFieldLen`

Maximum length of a field name of parsed bodies.

Type: integer greater than 0 &mdash; Default: `16384`

## `$maxInputVars`

Maximum number of input vars (in query string or parsed body).

Type: integer greater or equal to 0 &mdash; Default: `200`

## `$maxBodySize`

Default maximum size of HTTP bodies. [Can be increased by calling [`HttpDriver::upgradeBodySize($ireq)`](httpdriver.md) or more commonly [`Response::getBody($size)`](response.md).]

Type: integer greater than or equal to 0 &mdash; Default: `131072` (128 KiB)

## `$maxHeaderSize`

Maximum header size of a HTTP request.

Type: integer greater than 0 &mdash; Default: `32768`

## `$ioGranularity`

Granularity at which reads from the socket and into the bodies are performed.

Type: integer greater than 0 &mdash; Default: `32768`

## `$softStreamCap`

Limit at which the internal buffers are considered saturated and resolution of `Promise`s returned by `Response::write()` is delayed until the buffer sizes fall below it.

Type: integer greater than or equal to 0 &mdash; Default: `131072`

## `$deflateMinimumLength`

Minimum length before any compression is applied.

Type: integer greater than or equal to 0 &mdash; Default: `860`

## `$deflateBufferSize`

Buffer size before data is compressed (except it is ended or flushed before).

Type: integer greater than 0 &mdash; Default: `8192`

## `$chunkBufferSize`

Buffer size before data is being chunked (except it is ended or flushed before).

Type: integer greater than 0 &mdash; Default: `8192`

## `$outputBufferSize`

Buffer size before data is written onto the stream (except it is ended or flush before).

Type: integer greater than 0 &mdash; Default: `8192`

## `$shutdownTimeout`

Milliseconds before the Server is forcefully shut down after graceful stopping has been initiated.

Type: integer greater or equal to 0 &mdash; Default: `3000`


CHANGELOG.md
### 0.7.4

 - Fixed an issue where the timing of WebSocket writes could cause out of order messages.

### 0.7.3

 - Allow `amphp/file ^0.3`
 - Fixed sending pending responses when shutting down the server (#211)
 - More portable CPU count detection (#207)
 - WebSocket updates:
    - Reject websocket frames if RSV is not equal to `0` (will update as extensions are supported).
    - Accept zero-length frames starting a message.
    - Reject continuations if there's no started message.
    - Disable streaming for frames. A single frame is now always buffered. A message can still be streamed via multiple frames.

### 0.7.2

 - Fixed reading request port with HTTP/2

### 0.7.1

 - Fixed connection hangs if the process is forked while serving a request. See #182 and #192.

### 0.7.0

 - Fixed incorrect log level warning.
 - Fixed issue with referenced IPC socket blocking server shutdown.
 - Added support for unix sockets.
 - Added support for wildcard server names such as `localhost:*`, `*:80` and `*:*`.
 - Fixed buggy HTTP/1 pipelining.
 - Handle promises returned from generators in the config file correctly.
 - Added `-u` / `--user` command line option.
 - Correctly decode URL parameters with `urldecode()` instead of `rawurldecode()`.
 - Fixed freeze of websocket reading with very low `maxFramesPerSecond` and `maxBytesPerMinute`.
 - Removed `Router::__call()` magic, use `Router::route()` instead.

### 0.6.2

Retag of `v0.6.0`, as `v0.6.1` has been tagged wrongly, which should have been `v0.7.0`.

### 0.6.1

**Borked release.** Should have been `v0.7.0` and has been tagged as `v0.7.0` now.

### 0.6.0

Initial release based on Amp v2.

- Config files must return an instance of `Host` or an array of `Host` instances.
- `Aerys\Response::stream()` renamed to `write()` so that `Aerys\Response` may implement `Amp\ByteStream\OutputStream`.
- `Aerys\WebSocket\Endpoint::send()` split into three methods: `send()`, `broadcast()`, and `multicast()`.
- `Aerys\Body` removed. `Request::getBody()` returns an instance of `Amp\ByteStream\Message`.

### 0.5.0

_No information available._

### 0.4.7

_No information available._

### 0.4.6

_No information available._

### 0.4.5

_No information available._

### 0.4.4

_No information available._

### 0.4.3

 - Implemented monitoring system.
 - Always properly close HTTP/1.0 connections.
 - Fixed wildcard address matching.

### 0.4.2

_No information available._

### 0.4.1

 - Fixed caching issues in `Router` if multiple methods exist for specific URIs.
 
### 0.4.0

_No information available._

### 0.3.0

_No information available._

### 0.2.0

_No information available._

### 0.1.0

_No information available._


examples/options.php
<?php

/**
 * Aerys reserves certain configuration keys for internal use:
 * 
 * - "aerys.options"
 * - "aerys.beforeStart"
 * - "aerys.definitions"
 * 
 * All other keys are considered host blocks. Applications should avoid prefixing their host block
 * keys with "aerys.*" to avoid potential conflicts with future reserved keys. Otherwise,
 * applications may specify any value they like to internally differentiate individual hosts in the
 * config array.
 * 
 * We can assign server-wide configuration directives using the "aerys.options" key as demonstrated
 * below.
 */

$myApp = function(array $asgiEnv) {
    $body = '<html><body><h1>Hello, world.</h1></body></html>';
    return [$status = 200, $reason = 'OK', $headers = [], $body];
};

$config = [
    'my-site' => [ // <-- All unreserved keys are treated as host definitions
        'listenOn' => '*:1337',
        'application' => $myApp
    ],
    'aerys.options' => [ // <-- Reserved key used to set server options
        'logErrorsTo'           => 'php://stderr',
        'maxConnections'        => 2500,
        'maxRequests'           => 150,
        'keepAliveTimeout'      => 5,
        'disableKeepAlive'      => FALSE,
        'maxHeaderBytes'        => 8192,
        'maxBodyBytes'          => 10485760,
        'defaultContentType'    => 'text/html',
        'defaultTextCharset'    => 'utf-8',
        'sendServerToken'       => FALSE,
        'normalizeMethodCase'   => TRUE,
        'autoReasonPhrase'      => TRUE,
        'requireBodyLength'     => TRUE,
        'allowedMethods'        => ['GET', 'HEAD', 'OPTIONS', 'POST', 'PUT', 'PATCH', 'TRACE', 'DELETE'],
        'socketSoLingerZero'    => FALSE, // Requires PHP's ext/sockets extension
        'verbosity'             => 1,     // Server::SILENT (0), Server::QUIET (1), Server::LOUD (2)
        'defaultHost'           => NULL   // Must match a registered Host ID, e.g. mysite.com:80 or *:1337
    ]
];
